Proposed communication to community for OCP 4.9 upgrade and API concerns (#166)
BCDevOps/platform-services-status-page-notifications: notifications/Upgrading-OCP49-API.md

> # API Versioning Issues to Note Before We Upgrade from Openshift 4.8 to 4.9

Reviewer: ```suggestion

# What is happening?

Platform Services Team is preparing for Openshif 4.9 upgrade in June 2022. There are significant differences between API Versioning in Openshift 4.8 that is intalled on the Platform today and Openshift 4.9. 
```

Reviewer 2: Can we leave out any specific dates for the upgrade? It may be before June if things line up for us.

## Introduction - Openshift Upgrades and Kubernetes

Red Hat vendor link: <https://docs.openshift.com/container-platform/4.9/release_notes/ocp-4-9-release-notes.html#ocp-4-9-removed-kube-1-22-apis><br>
Kubernetes vendor link: <https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-22>

Red Hat Openshift bases itself off of Kubernetes and then adds additional technologies to create the product we all use today. This means that for each version of Openshift, it is matched against a specific version of Kubernetes.

> For example, The version of Kubernetes that is used when we upgraded to Openshift 4.8 is Kubernetes version 1.21. We know this in advance since Red Hat in their release notes will put this version change details in the notes, since many technologies out there might not necessarily have compatibility details regarding Openshift versions, but do have this information for Kubernetes versions. Click [here](https://docs.openshift.com/container-platform/4.8/release_notes/ocp-4-8-release-notes.html#ocp-4-8-about-this-release) to see this for Openshift 4.8, which runs with Kubernetes version 1.21.

Reviewer: ```suggestion
For example, Openshift 4.8 which the Platform was upgraded to in winter 2022 uses Kubernetes version 1.21. Red Hat usually includes the Kubernetes version used by each Openshift version in the release notes, as many technologies out there might not necessarily include details about their compatibility with Openshift versions, but typically have this information for Kubernetes versions. Click [here](https://docs.openshift.com/container-platform/4.8/release_notes/ocp-4-8-release-notes.html#ocp-4-8-about-this-release) to see this for Openshift 4.8, which runs with Kubernetes version 1.21.
```

## API Versioning

<https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-groups-and-versioning>

When we create or update Kubernetes resources (ie: Deployments, Routes, RoleBindings) inside Openshift, these resources will be defined with various pieces of information, including an API version. This information is processed and handled by the API Server accordingly.

As an API resource matures over time in terms of how it is implemented and maintained, the API maintainers will choose to update the API version to match its level of maturity. For the current wave of API upgrades that will take place when we upgrade Openshift from 4.8 to 4.9 (thus upgrade from Kubernetes 1.21 to 1.22), the affected API resources will be all easily identifiable as being an API resource with version `v1beta1` and that version in Openshift 4.9 will need to be just `v1`.

> 

Reviewer: ```suggestion
# What impact the upgrade to Openshift 4.8 may have on my app?
```

Reviewer 2: ```suggestion
# What impact may the upgrade to Openshift 4.9 have on my app?
```

## Which API Resources are Affected

<https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-22>

There are a number of Kubernetes resources affected by this upcoming upgrade, but most of them do not apply to people performing development work for applications hosted inside Openshift. The applicable resources can be broken down into three specific Kubernetes resources.

### Ingress
<https://kubernetes.io/docs/reference/using-api/deprecation-guide/#ingress-v122>

- API groups `networking.k8s.io` and `extensions` are to be replaced with just `networking.k8s.io`.
- API version changes from `v1beta1` to `v1`.
- Above link details changes regarding how the Ingress contents are defined.

### Role and RoleBinding
<https://kubernetes.io/docs/reference/using-api/deprecation-guide/#rbac-resources-v122>

- API version changes from `v1beta1` to `v1`.
- no other changes to note.

> 

Reviewer 3: Seems like we should include CustomResourceDefinition in this list, as v1beta1 will have to be replaced with v1.
https://kubernetes.io/docs/reference/using-api/deprecation-guide/#customresourcedefinition-v122

Reviewer 2: Product teams don't have the permissions to create CRDs/ Only cluster-admins can. All the operators AdvSol manages are already updated, or will update as part of the cluster upgrade.

Are the Platform Services owned operators updated? If not, we'll find out during the lab testing and roll out a CCM fix as part of the upgrade. Should still be transparent to Product Teams.

Reviewer 3: There are kustomization.yaml files that reference a v1beta1 API version, but that is consistent with current Kubernetes docs and not listed in the Kubernetes API upgrade list.
`apiVersion: kustomize.config.k8s.io/v1beta1`
https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/

This method of auditing will also generate false positives, since we may see LIST requests involving a deprecated API, but that's more due to the API server just picking the first API version available to it. The bigger concern during this audit is identifying deprecated API calls that create or change the specific resources we want to watch out for. When we upgrade to Openshift 4.9, those deprecated API versions will no longer exist, and LIST requests and similar methods will just use the v1 API version.

### Request Test Environment in a LAB cluster

> For important applications needing full assurance that nothing will break before we move forward to upgrade our PROD Openshift clusters, the best solution if you are not certain that 100% of the issues are addressed is to request via regular channels for requesting a new namespace and specify that LAB is required and why (Openshift 4.9 compatibility tests). We do upgrade our LAB clusters to the new version of Openshift first for both internal testing as well as for scenarios like this.

Reviewer: ```suggestion
For important applications needing full assurance that nothing will break before we move forward to upgrade our PROD Openshift clusters to Openshift 4.9, the recommendation is to request Platform Services Team (mailto:PlatformServicesTeam@gov.bc.ca)  a new project set in our LAB  cluster where you can deploy your app and test it with the Openshift 4.9 before the upgrade takes place in the production clusters. We do upgrade our LAB clusters to the new version of Openshift first for both internal testing as well as for scenarios like this.
```
